src/Microsoft.AspNetCore.Mvc.Razor/TagHelpers/TagHelperComponentTagHelper.cs

```diff
@@ -49,11 +50,62 @@ public abstract class TagHelperComponentTagHelper : TagHelper
             _logger = loggerFactory.CreateLogger(GetType());
         }
 
+        /// <summary>
+        /// Creates a new <see cref="TagHelperComponentTagHelper"/> and orders the 
+        /// the collection of <see cref="ITagHelperComponent"/>s in <see cref="ITagHelperComponentManager.Components"/>.
+        /// </summary>
+        /// <param name="manager">The <see cref="ITagHelperComponentManager"/> which contains the collection
+        /// of <see cref="ITagHelperComponent"/>s.</param>
+        /// <param name="propertyActivator">The <see cref="ITagHelperComponentPropertyActivator"/> activates
+        /// properties of <see cref="ITagHelperComponent"/>s.</param>
+        /// <param name="loggerFactory">The <see cref="ILoggerFactory"/>.</param>
+        /// <remarks>The <see cref="ITagHelperComponentManager.Components"/> are ordered after the 
+        /// creation of the <see cref="ITagHelperComponentManager"/> to position the <see cref="ITagHelperComponent"/>s
+        /// added from controllers and views correctly.</remarks>
+        public TagHelperComponentTagHelper(
+            ITagHelperComponentManager manager,
+            ITagHelperComponentPropertyActivator propertyActivator,
+            ILoggerFactory loggerFactory)
+        {
+            if (manager == null)
+            {
+                throw new ArgumentNullException(nameof(manager));
+            }
+
+            if (propertyActivator == null)
+            {
+                throw new ArgumentNullException(nameof(propertyActivator));
+            }
+
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _components = manager.Components.OrderBy(p => p.Order).ToArray();
+            PropertyActivator = propertyActivator;
+            _logger = loggerFactory.CreateLogger(GetType());
+        }
+
+        public ITagHelperComponentPropertyActivator PropertyActivator;
+
+        [ViewContext]
+        [HtmlAttributeNotBound]
+        public ViewContext ViewContext { get; set; }
+
         /// <inheritdoc />
         public override void Init(TagHelperContext context)
         {
+            if (PropertyActivator == null)
+            {
+                var serviceProvider = ViewContext.HttpContext.RequestServices;
+                PropertyActivator = serviceProvider.GetService(typeof(ITagHelperComponentPropertyActivator))
+                    as TagHelperComponentPropertyActivator;
+            }
```

Comment:
@rynowak I know you suggested using `TagHelperInitializer` but just curious if this is enough/acceptable instead. If not I can look into using `TagHelperInitializer`

Reply:
1. If we go with this approach, use `GetRequiredService<ITagHelperComponentPropertyActivator>()`
2. If you need the concrete `TagHelperComponentPropertyActivator` type, something is wrong
3. I lean toward preferring an `ITagHelperInitializer<TTagHelper>` for each known `TagHelperComponentTagHelper` subclasses but could be convinced otherwise if it's a lot of code to register initializers by default

Reply:
If we go with the initializers approach, will someone writing a new subclass for `TagHelperComponentTagHelper` need to write their own intializer? @dougbu

Reply:
Yes

Reply:
Okay then I don't feel very strongly about using initializers because right now it is very easy to add new TagHelpers which look exactly like Head and Body but target a different html element (like footer etc.). If we add another requirement, it may not be as extensible for some users. cc @rynowak @Eilon Thoughts?